reactive/calico.py

```diff
@@ -21,19 +26,96 @@
 ETCD_CERT_PATH = os.path.join(CALICOCTL_PATH, 'etcd-cert')
 ETCD_CA_PATH = os.path.join(CALICOCTL_PATH, 'etcd-ca')
 CALICO_CIDR = '192.168.0.0/16'
+CALICO_UPGRADE_DIR = '/opt/calico-upgrade'
 
 
 @hook('upgrade-charm')
 def upgrade_charm():
     remove_state('calico.binaries.installed')
     remove_state('calico.cni.configured')
+    remove_state('calico.service.installed')
+    remove_state('calico.pool.configured')
+    remove_state('calico.cni.configured')
+    remove_state('calico.npc.deployed')
     try:
         log('Deleting /etc/cni/net.d/10-calico.conf')
         os.remove('/etc/cni/net.d/10-calico.conf')
     except FileNotFoundError as e:
         log(e)
+    if is_leader() and not leader_get('calico-v3-data-ready'):
+        leader_set({
+            'calico-v3-data-migration-needed': True,
+            'calico-v3-npc-cleanup-needed': True,
+            'calico-v3-completion-needed': True
+        })
+
+
+@when('leadership.is_leader', 'leadership.set.calico-v3-data-migration-needed',
+      'etcd.available', 'calico.etcd-credentials.installed')
+def upgrade_v3_migrate_data():
+    status_set('maintenance', 'Migrating data to Calico 3')
+    try:
+        calico_upgrade.configure()
+        calico_upgrade.dry_run()
+        calico_upgrade.start()
+    except Exception:
+        log(traceback.format_exc())
+        message = 'Calico upgrade failed, see debug log'
+        status_set('blocked', message)
+        return
+    leader_set({'calico-v3-data-migration-needed': None})
```

Comment:
So when the upgrade fails we set this to None. How would the user recover from this?